### PR TITLE
viewport-addon: Make the addon configurable

### PR DESCRIPTION
**Pull request by mshaaban088**
Note: This is my first open-source PR.

### Edit
I addressed #2395 and the respective [comment](https://github.com/storybooks/storybook/issues/2395#issuecomment-369578940)

Issue: Currently you cannot customize or add new viewports/devices as described in the README file

## What I did
* Exported a `configure` function to easily customize the addon
  * Add new device/viewport
  * Replace the entire viewports set
   * Change the default viewport for all stories
* Exported a decorator function and wrapper component to change the default viewport per story
* Replaced the `default` option with a new viewport called `responsive` to treat the default as 1st-class citizen
* Added 2 new devices, and different stories for the`viewport` addon to the `official-storybook` example
* Organized the code respectively

### Add new device
```javascript
// config.js
import { configure, INITIAL_VIEWPORTS } from '@storybook/addon-viewport';

const extraViewports = {
  kindleFire2: {
    name: 'Kindle Fire 2',
    styles: {
      width: '600px',
      height: '963px'
    }
  },
  kindleFireHD: {
    name: 'Kindle Fire HD',
    styles: {
      width: '533px',
      height: '801px'
    }
  }
};

configure({
  viewports: {
    ...INITIAL_VIEWPORTS,
    ...extraViewports
  }
});
```
add

### Use new list of devices
```javascript
// config.js

import { configure } from '@storybook/addon-viewport';

const newViewports = {
  kindleFire2: {
    name: 'Kindle Fire 2',
    styles: {
      width: '600px',
      height: '963px'
    }
  },
  kindleFireHD: {
    name: 'Kindle Fire HD',
    styles: {
      width: '533px',
      height: '801px'
    }
  }
};

configure({
  viewports: newViewports
});
```

replace

### Change the default viewport

```javascript
// config.js

import { configure } from '@storybook/addon-viewport';

configure({
  defaultViewport: 'iphone6'
});
```

default-for-all

### Change the default viewport per story

```javascript
// file.stories.js

import React from 'react';
import { storiesOf } from '@storybook/react';
import { withViewport, Viewport } from '@storybook/addon-viewport';

storiesOf('Kindle Fire 2', module)
  .addDecorator(withViewport('kindleFire2'))
  .add('Inherited', () => (
    <div>
      I will be displayed on <b>Kindle Fire 2</b> viewport by default.
    </div>
  ))
  .add('Overridden', () => (
    <Viewport name="iphone6">
      <div>
        I will be displayed on <b>iPhone 6</b> viewport by default.
      </div>
    </Viewport>
  ));
```

## How to test
1. `cd examples/official-storybook`
2. `yarn storybook`
3. You will see 2 new devices were added to the `viewport` panel
4. Change the `extra-viewports.json` file
5. Observe the changes

Is this testable with jest or storyshots?
Yes

Does this need a new example in the kitchen sink apps?
No, I added it to the `official-storybook` already

Does this need an update to the documentation?
Yes

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: examples/cra-kitchen-sink/.storybook/config.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import { configure, setAddon } from '@storybook/react';
 import { setOptions } from '@storybook/addon-options';
+import { setViewports } from '@storybook/addon-viewport';
 
 // deprecated usage of infoAddon
 import infoAddon from '@storybook/addon-info';
 
+import viewports from './viewports.json';
```
Comment (igor-dv) [inline examples/cra-kitchen-sink/.storybook/config.js]: I think that you need to move the example to the `official-storybook`. @Hypnosphi  ?

File: addons/viewport/src/preview/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import addons from '@storybook/addons';
+import { ADD_VIEWPORTS_EVENT_ID, SET_VIEWPORTS_EVENT_ID } from '../shared';
+
+export function init() {}
```
Comment (igor-dv) [inline addons/viewport/src/preview/index.js]: What is this for?

**Review by igor-dv — COMMENTED** [review]
Congrats on your first PR! Great work.

File: addons/viewport/src/preview/index.js (hunk 2 above)
Comment (mshaaban088) [reply addons/viewport/src/preview/index.js]: I couldn't find a clear API documentation describing the preview API's, so I decided to follow the same approach as manager

Comment (igor-dv) [reply addons/viewport/src/preview/index.js]: Got it, but this method is a bit misleading, so let's remove it.

File: examples/cra-kitchen-sink/.storybook/config.js (hunk 1 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/.storybook/config.js]: Yes, that would be right

Comment (mshaaban088) [reply examples/cra-kitchen-sink/.storybook/config.js]: @igor-dv @Hypnosphi I moved it to `official-storybook`

File: addons/viewport/src/preview/index.js (hunk 2 above)
Comment (mshaaban088) [reply addons/viewport/src/preview/index.js]: I removed it

**Comment by mshaaban088** [comment]
@igor-dv @Hypnosphi thanks for reviewing it

File: addons/viewport/src/manager/components/Panel.js [hunk 3]
```diff
@@ -26,20 +47,42 @@ export class Panel extends Component {
...
+
+    this.setState({ viewports: newViewports });
+  };
+
+  addViewports = viewports => {
+    const newViewports = viewportsTransformer(viewports);
+
+    this.setState({
+      viewports: {
+        ...initialViewports,
```
Comment (Hypnosphi) [inline addons/viewport/src/manager/components/Panel.js]: Should it be current viewports instead?

**Review by Hypnosphi — CHANGES_REQUESTED** [review]
Please update addon documentation

File: addons/viewport/src/manager/components/Panel.js (hunk 3 above)
Comment (mshaaban088) [reply addons/viewport/src/manager/components/Panel.js]: My intention was to make `add` works by adding new viewports to the initial ones, but indeed your point makes more sense

Comment (mshaaban088) [reply addons/viewport/src/manager/components/Panel.js]: done

**Comment by mshaaban088** [comment]
@Hypnosphi I updated the documentation and addressed your comment regarding using previous `viewports` instead of `initialViewports`

**Review by Hypnosphi — APPROVED** [review]

**Comment by igor-dv** [comment]
@shilman, are we still merging the features? You wanted to lock in favor of the release.

**Comment by Hypnosphi** [comment]
@shilman if we're feature-frozen, we probably should edit those lines on master: https://github.com/storybooks/storybook/blob/master/dangerfile.js#L13-L15

**Review by danielduan — APPROVED** [review]

**Comment by mshaaban088** [comment]
I'm going to rebase with master again to not have those `Merge branch master into master`

**Comment by Hypnosphi** [comment]
> to not have those `Merge branch master into master`

We're actually OK with those

**Review by ndelangen — APPROVED** [review]

**Comment by mshaaban088** [comment]
@igor-dv @Hypnosphi @danielduan @ndelangen Could you please re-review it, apart from the conflicts, do you think this way of implementing it is better?
I really appreciate your time reviewing the initial one, but wanted to make everybody happy

**Review by Hypnosphi — APPROVED** [review]
LGTM

**Comment by Hypnosphi** [comment]
Please fix eslint warnings and unit tests

**Comment by mshaaban088** [comment]
@Hypnosphi Are you aware of any historical reason behind making `viewports` an object instead of array?
CC: @saponifi3d

**Comment by Hypnosphi** [comment]
No I'm not. I think we're free to change it in this PR as it wasn't exposed to user before. On the other side, it may be convenient as it is now, if user wants to pick only particular viewports from defaults or override some values in a particular viewport:

```
const { iphone6, ipad } = INITIAL_VIEWPORTS;

configure({
  viewports: {
    iphone6,
    ipad,
    ...newViewports
  }
});
```

```
configure({
  viewports: {
    ...INITIAL_VIEWPORTS,
    ipad: {
      ...INITIAL_VIEWPORTS.ipad,
      ...styles: {
        ...INITIAL_VIEWPORTS.ipad.styles,
        height: '1000px'
      }
    },
    ...newViewports
  }
});
```

**Comment by mshaaban088** [comment]
@Hypnosphi This is absolutely a valid use-case.

The reason why this question is popped up, is that the order of `Object.keys` is not guaranteed, hence, we cannot say we will pick the first one as default which makes the tests flickering as well.

**Comment by Hypnosphi** [comment]
@mshaaban088 can you please resolve the conflict?

**Comment by mshaaban088** [comment]
@Hypnosphi sure

**Comment by mshaaban088** [comment]
@Hypnosphi done

**Pull request merged**